File: source/tuxdb-shell/collection.py
from connection import connection
from env import Env


class Collection:

    __CORE_COMMAND = "collection"

    @staticmethod
    def help():
        helpString = ""

        helpString += "\n-- Collection functions --\n\n"
        helpString += "db.get(<collection name : String>).setCollectionName(<new collection name : String>) : This process sets the collection name\n"
        helpString += "db.get(<collection name : String>).getAllObjects() : This process shows all the objects in the collection\n"
        helpString += "db.get(<collection name : String>).findFromObjectId(<object id : String>) : This process finds a object associated with the object id in the collection\n"
        helpString += "db.get(<collection name : String>).find(<query : JSON>) : This process find the objects associated with the query in the collection \n"
        helpString += "db.get(<collection name : String>).insert(<data : JSON>) : This process inserts a object into the collection\n"
        helpString += "db.get(<collection name : String>).updateFromObjectId(<object id : String>, <update data : JSON>) : This process updates a object associated with the object id in the collection\n"
        helpString += "db.get(<collection name : String>).update(<query : JSON>, <update data : JSON>) : This process updates the object associated with the query in the collection\n"
        helpString += "db.get(<collection name : String>).deleteFromObjectId(<object id : String>) : This process deletes a object associated with the object id in the collection\n"
        helpString += "db.get(<collection name : String>).delete(<query : JSON>) : This process deletes a object associated with the query in the collection\n"
        helpString += "db.get(<collection name : String>).drop() : This process delete the collection from the currently used database\n"
        
        return helpString

    def __init__(self, collectionName: str):
        self.__collectionName = collectionName

    def setCollectionName(self, newCollectionName: str):
        if "/" in newCollectionName:
            return "New collection name doesn't contain this symbol '/'"

        head = [Collection.__CORE_COMMAND, "setCollectionName"]
        body = [self.__collectionName, newCollectionName]

        connection.send(head, body)

        return connection.read()

    def getAllObjects(self):
        head = [Collection.__CORE_COMMAND, "getAllObjects"]
        body = [self.__collectionName]

        connection.send(head, body)

        return connection.read()

    def findFromObjectId(self, objectId: str):
        head = [Collection.__CORE_COMMAND, "findFromObjectId"]
        body = [self.__collectionName, objectId]

        connection.send(head, body)

        return connection.read()

    def find(self, query: dict):
        head = [Collection.__CORE_COMMAND, "find"]
        body = [self.__collectionName, query]

        connection.send(head, body)

        return connection.read()

    def insert(self, data: dict):
        head = [Collection.__CORE_COMMAND, "insert"]
        body = [self.__collectionName, data]

        connection.send(head, body)

        return connection.read()

    def updateFromObjectId(self, objectId: str, updateData: dict):
        head = [Collection.__CORE_COMMAND, "updateFromObjectId"]
        body = [self.__collectionName, objectId, updateData]

        connection.send(head, body)

        return connection.read()

    def update(self, query: dict, updateData: dict):
        head = [Collection.__CORE_COMMAND, "update"]
        body = [self.__collectionName, query, updateData]

        connection.send(head, body)

        return connection.read()

    def deleteFromObjectId(self, objectId: str):
        head = [Collection.__CORE_COMMAND, "deleteFromObjectId"]
        body = [self.__collectionName, objectId]

        connection.send(head, body)

        return connection.read()

    def delete(self, query: dict):
        head = [Collection.__CORE_COMMAND, "delete"]
        body = [self.__collectionName, query]

        connection.send(head, body)

        return connection.read()

    def drop(self):
        head = [Collection.__CORE_COMMAND, "dropCollection"]
        body = [self.__collectionName]

        connection.send(head, body)

        return connection.read()

    def __str__(self) -> str:
        return f"{Env.database}.{self.__collectionName}"

File: api/tuxjava/src/main/java/io/github/burakpadr/tuxjava/Collection.java
package io.github.burakpadr.tuxjava;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import com.fasterxml.jackson.databind.ObjectMapper;

public class Collection {

    private final static String SERVICE_NAME = "collection";

    private Client client;
    private String databaseName;
    private String collectionName;

    public Collection(Client client, String databaseName, String collectionName) throws Exception {
        this.client = client;
        this.databaseName = databaseName;
        this.collectionName = collectionName;

        Map<String, Object> parameters = new HashMap<>();

        parameters.put("databaseName", databaseName);
        parameters.put("collectionName", collectionName);

        boolean collectionIsExist = (boolean) new ObjectMapper().readValue(
                client.send(SERVICE_NAME, "isExist", parameters), Boolean.class);

        if (!collectionIsExist)
            client.send(SERVICE_NAME, "createCollection", parameters);
    }

    @SuppressWarnings("unchecked")
    public Map<String, Object> setCollectionName(String newCollectionName) throws Exception {
        Map<String, Object> parameters = new HashMap<>();

        parameters.put("databaseName", databaseName);
        parameters.put("collectionName", collectionName);
        parameters.put("newCollectionName", newCollectionName);

        Map<String, Object> response = new HashMap<>();

        response = (Map<String, Object>) new ObjectMapper().readValue(client.send(SERVICE_NAME, "setCollectionName",
                parameters), response.getClass());

        if ((int) response.get("success") == 1)
            collectionName = newCollectionName;

        return response;
    }

    @SuppressWarnings("unchecked")
    public List<Map<String, Object>> getAllObjects() throws Exception {
        Map<String, Object> parameters = new HashMap<>();

        parameters.put("databaseName", databaseName);
        parameters.put("collectionName", collectionName);

        List<Map<String, Object>> response = new ArrayList<>();

        response = (List<Map<String, Object>>) new ObjectMapper().readValue(
                client.send(SERVICE_NAME, "getAllObjects", parameters),
                response.getClass());

        return response;
    }

    @SuppressWarnings("unchecked")
    public Map<String, Object> findFromObjectId(String objectId) throws Exception {
        Map<String, Object> parameters = new HashMap<>();

        parameters.put("databaseName", databaseName);
        parameters.put("collectionName", collectionName);
        parameters.put("objectId", objectId);

        Map<String, Object> response = new HashMap<>();

        response = (Map<String, Object>) new ObjectMapper().readValue(
                client.send(SERVICE_NAME, "findFromObjectId", parameters),
                response.getClass());

        return response;
    }

    @SuppressWarnings("unchecked")
    public List<Map<String, Object>> find(Map<String, Object> query) throws Exception {
        Map<String, Object> parameters = new HashMap<>();

        parameters.put("databaseName", databaseName);
        parameters.put("collectionName", collectionName);
        parameters.put("query", query);

        List<Map<String, Object>> response = new ArrayList<>();

        response = (List<Map<String, Object>>) new ObjectMapper().readValue(
                client.send(SERVICE_NAME, "find", parameters),
                response.getClass());

        return response;
    }

    @SuppressWarnings("unchecked")
    public Map<String, Object> insert(Map<String, Object> data) throws Exception {
        Map<String, Object> parameters = new HashMap<>();

        parameters.put("databaseName", databaseName);
        parameters.put("collectionName", collectionName);
        parameters.put("data", data);

        Map<String, Object> response = new HashMap<>();

        response = (Map<String, Object>) new ObjectMapper().readValue(
                client.send(SERVICE_NAME, "insert", parameters),
                response.getClass());

        return response;
    }

    @SuppressWarnings("unchecked")
    public Map<String, Object> updateFromObjectId(String objectId, Map<String, Object> updateData) throws Exception {
        Map<String, Object> parameters = new HashMap<>();

        parameters.put("databaseName", databaseName);
        parameters.put("collectionName", collectionName);
        parameters.put("objectId", objectId);
        parameters.put("updateData", updateData);

        Map<String, Object> response = new HashMap<>();

        response = (Map<String, Object>) new ObjectMapper().readValue(client.send(SERVICE_NAME, "updateFromObjectId",
                parameters), response.getClass());

        return response;
    }

    @SuppressWarnings("unchecked")
    public Map<String, Object> update(Map<String, Object> query, Map<String, Object> updateData) throws Exception {
        Map<String, Object> parameters = new HashMap<>();

        parameters.put("databaseName", databaseName);
        parameters.put("collectionName", collectionName);
        parameters.put("query", query);
        parameters.put("updateData", updateData);

        Map<String, Object> response = new HashMap<>();

        response = (Map<String, Object>) new ObjectMapper().readValue(
                client.send(SERVICE_NAME, "update", parameters),
                response.getClass());

        return response;
    }

    @SuppressWarnings("unchecked")
    public Map<String, Object> deleteFromObjectId(String objectId) throws Exception {
        Map<String, Object> parameters = new HashMap<>();

        parameters.put("databaseName", databaseName);
        parameters.put("collectionName", collectionName);
        parameters.put("objectId", objectId);

        Map<String, Object> response = new HashMap<>();

        response = (Map<String, Object>) new ObjectMapper().readValue(
                client.send(SERVICE_NAME, "deleteFromObjectId", parameters),
                response.getClass());

        return response;
    }

    @SuppressWarnings("unchecked")
    public Map<String, Object> delete(Map<String, Object> query) throws Exception {
        Map<String, Object> parameters = new HashMap<>();

        parameters.put("databaseName", databaseName);
        parameters.put("collectionName", collectionName);
        parameters.put("query", query);

        Map<String, Object> response = new HashMap<>();

        response = (Map<String, Object>) new ObjectMapper().readValue(
                client.send(SERVICE_NAME, "delete", parameters),
                response.getClass());

        return response;
    }

    @SuppressWarnings("unchecked")
    public Map<String, Object> drop() throws Exception {
        Map<String, Object> parameters = new HashMap<>();

        parameters.put("databaseName", databaseName);
        parameters.put("collectionName", collectionName);

        Map<String, Object> response = new HashMap<>();

        response = (Map<String, Object>) new ObjectMapper().readValue(
                client.send(SERVICE_NAME, "drop", parameters), response.getClass());

        return response;
    }

    @Override
    public String toString() {
        return String.format("Database Name -> %s\nCollection Name -> %s", databaseName, collectionName);
    }
}
